dynet/io.cc

```diff
@@ -0,0 +1,246 @@
+  this->deserialize(model, key);
+}
+
+void Pack::load(ParameterCollection & model,
+                const std::vector<std::string> & filter_lst,
+                const std::string & key) {
+  // TODO
+}
+
+bool Pack::duplicate_key_check(const std::string & key) {
+  std::ifstream f(fn_meta);
+  std::string line;
+  while (std::getline(f, line)) {
+    auto kv = dynet::str_split(line, ':');
+    if (kv[0] == key) return false;
+  }
+  f.close();
+  return true;
+}
+
+void Pack::serialize(const ParameterCollection & model, const std::string & key, bool is_append) {
+  std::ofstream os;
+  if (is_append) {
+    os.open(fn, std::ofstream::app);
+  } else {
+    os.open(fn);
+  }
+  os.seekp(this->offset);
+  os << '#' << std::endl; // identifier of beginning of the ParameterCollection
+  auto params = model.get_parameter_storages();
+  for (auto & param : params) {
+    os << "#Parameter#" << std::endl;
+    os << param->name << std::endl;
+    os << param->dim << std::endl;
+    os << param->values << std::endl;
+    os << param->g << std::endl;
+  }
+  auto lookup_params = model.get_lookup_parameter_storages();
+  for (auto & lookup_param: lookup_params) {
+    os << "#LookupParameter#" << std::endl;
+    os << lookup_param->name << std::endl;
+    os << lookup_param->all_dim << std::endl;
+    os << lookup_param->dim << std::endl;
+    os << lookup_param->all_values << std::endl;
+    os << lookup_param->all_grads << std::endl;
+  }
+  this->offset = os.tellp();
+  os.close();
+}
+
+void Pack::deserialize(ParameterCollection & model, const std::string & key) {
+  std::ifstream meta_f(fn_meta);
+  std::ifstream f(fn);
+  // find the offset of the key
+  long long local_offset = -1;
+
+  std::string line;
+  if (key.size() == 0) {
+    // case for no key specified
+    local_offset = 0;
+  } else {
+    while (std::getline(meta_f, line)) {
+      auto kv = dynet::str_split(line, ':');
+      if (kv[0] == key) {
+        local_offset = std::stoll(kv[1]);
+        break;
+      }
+    }
+  }
+  if (local_offset == -1) {
+    DYNET_RUNTIME_ERR("Load error: no such key: " + key);
+  }
+
+  // check identifier
+  f.seekg(local_offset);
+  std::getline(f, line);
+  if (line != "#") {
+    DYNET_RUNTIME_ERR("Invalid model file format. Check this line: " + line);
+  }
+
+  std::string ns = model.get_namespace();
+  // read parameters
+  std::getline(f, line);
+  while (line == "#Parameter#") {
+    std::getline(f, line);
+    if (dynet::startswith(line, ns) == false) {
+      DYNET_RUNTIME_ERR("Inconsistent namespace error: " + line + " | " + ns);
+    }
+    auto name = line;
+
+    Dim d;
+    std::getline(f, line);
+    std::istringstream iss(line);
+    iss >> d;
+
+    // add param into input model
+    Parameter param = model.add_parameters(d);
+    param.get_storage().name = name;
+
+    // read param.get_storage().values
+    std::vector<float> params_lst;
+    auto deserialize_tensor_lambda = [&] () {
```

Comment:
Maybe it's better to just define this and the following one as a private (inline?) member of Pack?